mypy_django_plugin/transformers/managers.py

```diff
@@ -183,112 +247,182 @@ def create_new_manager_class_from_from_queryset_method(ctx: DynamicClassDefConte
+    if manager_sym and isinstance(manager_sym.node, TypeInfo):
         # This is just a deferral run where our work is already finished
         return
 
     callee = ctx.call.callee
     assert isinstance(callee, MemberExpr)
     assert isinstance(callee.expr, RefExpr)
 
-    base_manager_info = callee.expr.node
-    if base_manager_info is None:
+    manager_base = callee.expr.node
+    if manager_base is None:
         if not semanal_api.final_iteration:
             semanal_api.defer()
         return
 
-    assert isinstance(base_manager_info, TypeInfo)
+    assert isinstance(manager_base, TypeInfo)
 
     passed_queryset = ctx.call.args[0]
     assert isinstance(passed_queryset, NameExpr)
 
-    derived_queryset_fullname = passed_queryset.fullname
-    if derived_queryset_fullname is None:
+    if passed_queryset.fullname is None:
         # In some cases, due to the way the semantic analyzer works, only passed_queryset.name is available.
         # But it should be analyzed again, so this isn't a problem.
         return
+    elif semanal_api.is_class_scope():
+        # We force `.from_queryset` to be called _outside_ of a class body. So we'll
+        # skip doing any work if we're inside of one..
+        return
 
-    base_manager_instance = fill_typevars(base_manager_info)
-    assert isinstance(base_manager_instance, Instance)
-    new_manager_info = semanal_api.basic_new_typeinfo(
-        ctx.name, basetype_or_fallback=base_manager_instance, line=ctx.call.line
-    )
-
-    sym = semanal_api.lookup_fully_qualified_or_none(derived_queryset_fullname)
-    assert sym is not None
-    if sym.node is None:
+    queryset_sym = semanal_api.lookup_fully_qualified_or_none(passed_queryset.fullname)
+    assert queryset_sym is not None
+    if queryset_sym.node is None:
         if not semanal_api.final_iteration:
             semanal_api.defer()
-        else:
-            # inherit from Any to prevent false-positives, if queryset class cannot be resolved
-            new_manager_info.fallback_to_any = True
         return
 
-    derived_queryset_info = sym.node
-    assert isinstance(derived_queryset_info, TypeInfo)
-
-    new_manager_info.line = ctx.call.line
-    new_manager_info.type_vars = base_manager_info.type_vars
-    new_manager_info.defn.type_vars = base_manager_info.defn.type_vars
-    new_manager_info.defn.line = ctx.call.line
-    new_manager_info.metaclass_type = new_manager_info.calculate_metaclass_type()
-    # Stash the queryset fullname which was passed to .from_queryset
-    # So that our 'resolve_manager_method' attribute hook can fetch the method from that QuerySet class
-    new_manager_info.metadata["django"] = {"from_queryset_manager": derived_queryset_fullname}
+    queryset_info = queryset_sym.node
+    assert isinstance(queryset_info, TypeInfo)
 
     if len(ctx.call.args) > 1:
         expr = ctx.call.args[1]
         assert isinstance(expr, StrExpr)
```

Comment:
Maybe an edge case, but it's also valid to explicitly call this with `None` or a dynamic string. And I don't like that we crash on type errors. Maybe just return if the type is not None or a string, mypy will emit a type error then anyway, and emit an error if it's not `None` or a literal string (as we're then unable to resolve the name)?

I guess that's not really related to this PR, but would be nice to fix this when you're rewriting this anyway

Reply:
Good point. I might as well update it here.

Reply:
I realised that a `TypeError` is raised from `builtins.type` if the name value in the 3-arg call isn't `str`. So I changed to fallback on the default value unless `class_name` is a hard coded `StrExpr`.